server/partitions.go

```diff
@@ -0,0 +1,343 @@
+		if _, err := p.s.nc.ChanSubscribe(pubSubject, p.msgsCh); err != nil {
+			return fmt.Errorf("could not subscribe to publish subject %q, %v", channelName, err)
+		}
+	}
+	// Add a dedicated subject for when we try to schedule a control
+	// message to be processed after all messages from a given client
+	// have been processed.
+	p.ctrlMsgSubject = fmt.Sprintf("%s.%s", p.s.info.Publish, nats.NewInbox())
+	if _, err := p.s.nc.ChanSubscribe(p.ctrlMsgSubject, p.msgsCh); err != nil {
+		return fmt.Errorf("could not subscribe to subject %q, %v", p.ctrlMsgSubject, err)
+	}
+	return nil
+}
+
+// Sends a request to the rest of the cluster and wait a bit for
+// responses (we don't know if or how many servers there may be).
+// No server lock used since this is called inside RunServerWithOpts().
+func (p *partitions) checkChannelsUniqueInCluster() error {
+	// We will use a subscription on an inbox to get the replies
+	replyInbox := nats.NewInbox()
+	replySub, err := p.nc.SubscribeSync(replyInbox)
+	if err != nil {
+		return fmt.Errorf("unable to create sync subscription: %v", err)
+	}
+	defer replySub.Unsubscribe()
+	// Send our list
+	if err := p.sendChannelsList(replyInbox); err != nil {
+		return fmt.Errorf("unable to send channels list: %v", err)
+	}
+	// Since we don't know how many servers are out there, keep
+	// calling NextMsg until we get a timeout
+	for {
+		reply, err := replySub.NextMsg(partitionsNextMsgTimeout)
+		if err == nats.ErrTimeout {
+			return nil
+		}
+		if err != nil {
+			return fmt.Errorf("unable to get partitioning reply: %v", err)
+		}
+		resp := spb.CtrlMsg{}
+		if err := resp.Unmarshal(reply.Data); err != nil {
+			return fmt.Errorf("unable to decode partitioning response: %v", err)
+		}
+		if len(resp.Data) > 0 {
+			return fmt.Errorf("channel %q causes conflict with channels on server %q",
+				string(resp.Data), resp.ServerID)
+		}
+	}
+}
+
+// Sends the list of channels to a known subject, possibly splitting the list
+// in several requests if it cannot fit in a single message.
+func (p *partitions) sendChannelsList(replyInbox string) error {
+	sendReq := func(channels []string) error {
+		buf := &bytes.Buffer{}
+		gob.NewEncoder(buf).Encode(channels)
+		req := &spb.CtrlMsg{
+			ServerID: p.s.serverID,
+			MsgType:  spb.CtrlMsg_Partitioning,
+			Data:     buf.Bytes(),
+		}
+		reqBytes, _ := req.Marshal()
+		return p.nc.PublishRequest(p.sendListSubject, replyInbox, reqBytes)
+	}
+	// Since the NATS message payload is limited, we need to repeat
+	// requests if all channels can't fit in a request.
+	maxPayload := p.nc.MaxPayload()
+	start := 0
+	end := 0
+	for end < len(p.channels) {
+		size := 0
+		tmpBuf := &bytes.Buffer{}
+		encoderTmpBuf := gob.NewEncoder(tmpBuf)
+		for end = start; end < len(p.channels); end++ {
+			encoderTmpBuf.Encode(p.channels[end])
+			size += tmpBuf.Len()
+			// Leave room for the CtrlMsg header
+			if size > int(maxPayload-50) {
```

Comment:
Changing the code a bit to create a request object with the server ID and type set and use `Size()` to get the encoded size that I use here instead of this value.